Added GHC plan for core

wduncanfraser opened the pull request:
Please see #615

This is dependent on #614 

Signed-off-by: W. Duncan Fraser <wdf@alasconnect.com>

thesentinels commented:
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

nellshamrell commented:
This is still blocked by #614

nellshamrell commented:
Hello!

I sincerely apologize for this being open for so long - we are doing a clean up as part of a new process to ensure we don't end up with PRs sitting this long again.  

Is this something you would like us to continue to move forward?  If yes, please let us know here, and I will leave this open and keep moving it forward.  If not, please go ahead and close the pull request.  

If I do not hear back within a week, I will go ahead and close the pull request, though you are welcome to re-open it if you'd like it to move forward in the future.

Thank you!

wduncanfraser commented:
@nellshamrell Yes I would like to move forward with this PR, though I think it would be best to move major compiler versions into their own namespaces. Due to library and code incompatibilities that aren't always trivial to resolve in large projects, it would probably make sense to namespace ghc80 vs ghc82, which I will be doing some work on over in https://github.com/alasconnect/habitat-plans and will then update this PR with those changes.

wduncanfraser commented:
@nellshamrell I've broken this out by major version. Any time the https://github.com/ghc/ghc/blob/master/rts/ClosureFlags.c updates in GHC, you aren't able to directly build new GHC versions with previous GHC, so I've re-bootstrapped when that has changed (GHC 7.10, GHC 8.2). This matches how nix bootstraps in GHC: https://github.com/NixOS/nixpkgs/tree/master/pkgs/development/compilers/ghc and gives us a reproducible build path form binary to source build.

I have left out a source built GHC 7.10 as I was having issues getting it to build, and have purely used the GHC 7.10 binary version to build GHC 8.0.

I've pulled in ncurses5-compat-libs as it is needed for the binary GHC bootstraps.

wduncanfraser commented:
I'm troubleshooting some issues that have cropped up post base-refresh.

wduncanfraser commented:
@nellshamrell Finally got a chance to work on this. I've updated everything to work post base refresh. 

It looks like the root cause was that the studio is running behind the base refresh, causing rpath/linker errors due to older binutils/glibc/etc. 

RPATH was getting set as 
```
 0x000000000000001d (RUNPATH)            Library runpath: [$ORIGIN/../haskeline-0.7.4.0:$ORIGIN/../ghc-8.2.2:$ORIGIN/../terminfo-0.4.1.0:$ORIGIN/../hoopl-3.10.2.2:$ORIGIN/../ghci-8.2.2:$ORIGIN/../ghc-boot-8.2.2:$ORIGIN/../transformers-0.5.2.0:$ORIGIN/../hpc-0.6.0.3:$ORIGIN/../template-haskell-2.12.0.0:$ORIGIN/../pretty-1.1.3.3:$ORIGIN/../ghc-boot-th-8.2.2:$ORIGIN/../process-1.6.1.0:$ORIGIN/../directory-1.3.0.2:$ORIGIN/../unix-2.7.2.2:$ORIGIN/../filepath-1.4.1.2:$ORIGIN/../time-1.8.0.2:$ORIGIN/../binary-0.8.5.1:$ORIGIN/../containers-0.5.10.2:$ORIGIN/../bytestring-0.10.8.2:$ORIGIN/../deepseq-1.4.3.0:$ORIGIN/../array-0.5.2.0:$ORIGIN/../base-4.10.1.0:$ORIGIN/../integer-gmp-1.0.1.0:$ORIGIN/../ghc-prim-0.5.1.1:$ORIGIN/../rts]
```

When it should look like

```
 0x000000000000001d (RUNPATH)            Library runpath: [$ORIGIN/../haskeline-0.7.4.0:$ORIGIN/../ghc-8.2.2:$ORIGIN/../terminfo-0.4.1.0:$ORIGIN/../hoopl-3.10.2.2:$ORIGIN/../ghci-8.2.2:$ORIGIN/../ghc-boot-8.2.2:$ORIGIN/../transformers-0.5.2.0:$ORIGIN/../hpc-0.6.0.3:$ORIGIN/../template-haskell-2.12.0.0:$ORIGIN/../pretty-1.1.3.3:$ORIGIN/../ghc-boot-th-8.2.2:$ORIGIN/../process-1.6.1.0:$ORIGIN/../directory-1.3.0.2:$ORIGIN/../unix-2.7.2.2:$ORIGIN/../filepath-1.4.1.2:$ORIGIN/../time-1.8.0.2:$ORIGIN/../binary-0.8.5.1:$ORIGIN/../containers-0.5.10.2:$ORIGIN/../bytestring-0.10.8.2:$ORIGIN/../deepseq-1.4.3.0:$ORIGIN/../array-0.5.2.0:$ORIGIN/../base-4.10.1.0:$ORIGIN/../integer-gmp-1.0.1.0:$ORIGIN/../ghc-prim-0.5.1.1:$ORIGIN/../rts:/hab/pkgs/alasconnect/ghc82/8.2.2/20180317061949/lib:/hab/pkgs/core/perl/5.22.1/20170513213942/lib:/hab/pkgs/core/gcc/5.2.0/20170513202244/lib:/hab/pkgs/core/glibc/2.22/20170513201042/lib:/hab/pkgs/core/gmp/6.1.0/20170513202112/lib:/hab/pkgs/core/libedit/3.1.20150325/20170514030323/lib:/hab/pkgs/core/libffi/3.2.1/20170514003538/lib:/hab/pkgs/core/libiconv/1.14/20170514021726/lib:/hab/pkgs/alasconnect/ncurses/6.0/20170607211541/lib]
```

I resolved this by bringing in core/binutils as a build dependency and overriding LD. This should also future proof the plan against base refreshes. 

I was able to get GHC 710 to build from source by using the embedded libffi, as it seems to have linker issues against newer versions of libffi.

I added generic `ghc` plan and updated versioned ghc plans for RFC 0003. I also added basic readme skeletons for all the plans.

Should I add myself as a code owner for the ghc plans?

cc @smacfarlane :trollface:

wduncanfraser commented:
Should also mention: Validated all builds. Used ghc 82 build out haskell tooling in https://github.com/alasconnect/habitat-plans/pull/18 and to build out all of our internal projects without issue.

nellshamrell commented:
Looks FANTASTIC!  Going to do one more check to see that these build locally, then will get these merged in an added to builder!  This is great work, ty!

nellshamrell commented:
All of these build successfully!  I'll merge this shortly, here's the order I'll be adding these to builder:

- [ ] ncurses5
- [ ] ghc710-bootstrap
- [ ] ghc710 (this will time out on builder - I'll be building this in a VM using the new versions of the above dependencies and then  will manually upload it to builder)
- [ ] ghc80 (this will time out on builder - I'll be building this in a VM using the new versions of the above dependencies and then will manually upload it to builder)
- [ ] ghc82-bootstrap
- [ ] ghc82 (this will time out on builder - I'll be building this in a VM using the new versions of the above dependencies and then will manually upload it to builder)
- [ ] ghc84 (this will time out on builder - I'll be building this in a VM using the new versions of the above dependencies and then will manually upload it to builder)
